modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java

```diff
@@ -0,0 +1,303 @@
+            } else {
+                Integer indexId0 = l0.get(i0);
+                Integer indexId1 = l1.get(i1);
+
+                int cmp = indexId0.compareTo(indexId1);
+
+                if (cmp > 0) {
+                    result.add(indexId1);
+                    i1++;
+                } else if (cmp < 0) {
+                    result.add(indexId0);
+                    i0++;
+                } else {
+                    result.add(indexId0);
+                    i0++;
+                    i1++;
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private void addListenersBusy() {
+        catalogService.listen(INDEX_STOPPING, (parameters, exception) -> {
+            if (exception != null) {
+                return failedFuture(exception);
+            }
+
+            return onIndexStopping((StoppingIndexEventParameters) parameters).thenApply(unused -> false);
+        });
+    }
+
+    private CompletableFuture<?> onIndexStopping(StoppingIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            ReadOnlyIndexInfo readOnlyIndexInfo = new ReadOnlyIndexInfo(
+                    parameters.tableId(),
+                    catalogActivationTimestampBusy(parameters.catalogVersion()),
+                    parameters.indexId()
+            );
+
+            readOnlyIndexes.add(readOnlyIndexInfo);
+
+            return nullCompletedFuture();
+        });
+    }
+
+    private long catalogActivationTimestampBusy(int catalogVersion) {
+        Catalog catalog = catalogService.catalog(catalogVersion);
+
+        assert catalog != null : catalogVersion;
+
+        return catalog.time();
+    }
+
+    private void recoveryReadOnlyIndexesBusy() {
+        int earliestCatalogVersion = catalogService.earliestCatalogVersion();
+        int latestCatalogVersion = catalogService.latestCatalogVersion();
+
+        // At the moment, we will only use tables from the latest version (not dropped), since so far only replicas for them are started
+        // on the node.
+        int[] tableIds = catalogService.tables(latestCatalogVersion).stream().mapToInt(CatalogObjectDescriptor::id).toArray();
+
+        Map<Integer, ReadOnlyIndexInfo> readOnlyIndexes = new HashMap<>();
+
+        for (int catalogVersion = earliestCatalogVersion; catalogVersion <= latestCatalogVersion; catalogVersion++) {
+            Catalog catalog = catalogService.catalog(catalogVersion);
+
+            assert catalog != null : catalogVersion;
+
+            for (int tableId : tableIds) {
+                for (CatalogIndexDescriptor index : catalog.indexes(tableId)) {
+                    if (index.status() == STOPPING) {
+                        readOnlyIndexes.computeIfAbsent(index.id(), indexId -> new ReadOnlyIndexInfo(tableId, catalog.time(), indexId));
+                    }
+                }
+            }
+        }
+
+        this.readOnlyIndexes.addAll(readOnlyIndexes.values());
```

Comment:
We also need RO indices from tables that are dropped, but not yet destroyed as for them we need to build indices on FST (because RO transactions in the past could read from them after FST finishes)

Reply:
fix it